stake-o-matic/src/validators_app.rs

```diff
@@ -0,0 +1,196 @@
+struct PingResponse {
+    answer: String,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct ValidatorsResponseEntry {
+    pub account: Option<String>,
+    pub active_stake: Option<u64>,
+    pub commission: Option<u8>,
+    pub created_at: Option<String>,
+    pub data_center_concentration_score: Option<i64>,
+    pub data_center_host: Option<String>,
+    pub data_center_key: Option<String>,
+    pub delinquent: Option<bool>,
+    pub details: Option<String>,
+    pub keybase_id: Option<String>,
+    pub name: Option<String>,
+    pub network: Option<String>,
+    pub ping_time: Option<f64>,
+    pub published_information_score: Option<i64>,
+    pub root_distance_score: Option<i64>,
+    pub security_report_score: Option<i64>,
+    pub skipped_slot_percent: Option<String>,
+    pub skipped_slot_score: Option<i64>,
+    pub skipped_slots: Option<u64>,
+    pub software_version: Option<String>,
+    pub software_version_score: Option<i64>,
+    pub stake_concentration_score: Option<i64>,
+    pub total_score: Option<i64>,
+    pub updated_at: Option<String>,
+    pub url: Option<String>,
+    pub vote_account: Option<String>,
+    pub vote_distance_score: Option<i64>,
+    pub www_url: Option<String>,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct ValidatorsResponse(Vec<ValidatorsResponseEntry>);
+
+impl AsRef<Vec<ValidatorsResponseEntry>> for ValidatorsResponse {
+    fn as_ref(&self) -> &Vec<ValidatorsResponseEntry> {
+        &self.0
+    }
+}
+
+#[allow(dead_code)]
+#[derive(Debug, Clone, Copy)]
+pub enum SortKind {
+    Score,
+    Name,
+    Stake,
+}
+
+impl std::fmt::Display for SortKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Self::Score => write!(f, "score"),
+            Self::Name => write!(f, "name"),
+            Self::Stake => write!(f, "stake"),
+        }
+    }
+}
+
+pub type Limit = u32;
+
+pub struct Client {
+    base_url: reqwest::Url,
+    cluster: ClusterJson,
+    api_token: String,
+    client: reqwest::blocking::Client,
+}
+
+impl Client {
+    pub fn new<T: AsRef<str>>(api_token: T) -> Self {
+        let config = ClientConfig {
+            api_token: api_token.as_ref().to_string(),
+            ..ClientConfig::default()
+        };
+        Self::new_with_config(config)
+    }
+
+    pub fn new_with_config(config: ClientConfig) -> Self {
+        let ClientConfig {
+            base_url,
+            cluster,
+            api_token,
+        } = config;
+        Self {
+            base_url: reqwest::Url::parse(&base_url).unwrap(),
+            cluster,
+            api_token,
+            client: reqwest::blocking::Client::new(),
+        }
+    }
+
+    fn request(
+        &self,
+        endpoint: Endpoint,
+        query: &HashMap<String, String>,
+    ) -> reqwest::Result<reqwest::blocking::Response> {
+        let url = self.base_url.join(&endpoint.path(&self.cluster)).unwrap();
+        let request = self
```

Comment:
Consider retry & timeout logic on the request in case of API errors. Also, think about a fail-safe action if the request results in an error. The validator locations do not change very often, so you might be able to use a cached copy of the last good API response.

Reply:
Thanks Brian!

> Consider retry & timeout logic on the request in case of API errors.

Totally!  I'll definitely implement some retry, if you ever make the API unstable :wink:  It's been 100% during dev. Bravo!

Odds are when we need it, I'll install retry logic up a level so as not to force any particular logic on all callers, assuming we'll have more users of this client eventually.

> Also, think about a fail-safe action if the request results in an error. The validator locations do not change very often, so you might be able to use a cached copy of the last good API response.

I took the easy route to fail-safe here and if any errors occur, fallback to an empty list (`or_default`) of violators.

https://github.com/solana-labs/solana/blob/c960344aad1c46eeb34e9b0779bc500d903a8622/stake-o-matic/src/main.rs#L1051-L1052

Thanks for reminding me that I _do_ want to add a log here so we know that something failed rather than just assume everyone's nice and distributed